DOC Improve performance section (#764)
silverstripe/developer-docs: en/02_Developer_Guides/08_Performance/06_ORM.md

>   check_and_repair_on_build: false

Author: I haven't renamed this to 05_ORM.md as github wanted to delete the old file and create a new file, rather than simply saying it's renamed. This would have made peer review very difficult, so I simply left a "gap" in the page numbering

Reviewer: Gaps are fine - arguably even desirable since it allows us to slot a new page in there in the future if we want.

  composite_db:
    Class: "DBClassNameVarchar('SilverStripe\\ORM\\DataObject', ['index' => false])"
```

> ## Eager loading {#eager-loading}

Reviewer: ```suggestion
## Eager loading
```
That anchor will be automatically generated. Adding it like this breaks direct-links in IDEs because they don't understand the syntax, so ideally we should only add an explicit anchor if it differs from what's auto-generated.

> Querying nested relationships inside a loop using the ORM is prone to the N + 1 query problem. To illustrate the N + 1 query problem, imagine a scenario where there are Teams with many child Players

Author: Moved from 02_Relations.md

This is a pretty large section so arguably it should have its own page

Reviewer: Yup, giving it its own page makes sense, please do that.

```php
use SilverStripe\ORM\DataList;

$teams = Team::get()->eagerLoad([
    'Players' => fn (DataList $list) => $list->filter(['Age:GreaterThan' => 18]),
    'Players.Fans' => fn (DataList $list) => $list->filter(['Name:PartialMatch:nocase' => 'Sam']),
    'Players.Fans.Events' => fn (DataList $list) => $list->filter(['DollarCost:LessThan' => 200]),
]);
```

If you have complex branching logic, and in some branches you want to avoid pre-filtering or perform different filtering, you can declare a different callback (or `null`) for that relation chain by calling `eagerLoad()` again. Note that subsequent callbacks *replace* previous callbacks.

```php
use SilverStripe\ORM\DataList;

$teams = Team::get()->eagerLoad([
    // Remove any callback that was previously defined for this relation chain
    'Players' => fn (DataList $list) => null,
    // Replace any callback that was previously defined for this relation chain.
    // If you want to apply *additional* filters rather than replacing existing ones, you will
    // need to declare the original filter again here.
    'Players.Fans.Events' => fn (DataList $list) => $list->filter(['DollarCost:GreaterThan' => 100]),
]);
```

> ## Making raw SQL queries {#raw-sql}

Author: Note that this section is new i.e. it's not part of the eager loading section that was moved

> If find the ORM is making needlessly innefficent SQL queries for a particular use case, then you can simply make raw SQL.

Reviewer: ```suggestion
If find the ORM is making needlessly inefficient SQL queries for a particular use case, then you can use raw SQL.
```

Author: Updated

> [!WARNING]
> Using raw SQL queries can make your code less more difficult to maintain. Only use raw SQL when the ORM is a clear bottleneck.

> ```php
> // app/src/Services/SomeService.php
> namespace App\Services;
> 
> use SilverStripe\ORM\DB;
> 
> class MyService
> {
>     public function makeRawQuery(): void
>     {
>         // You can use query() if using a fixed value
>         $result = DB::query('SELECT * FROM "MyTable" WHERE "MyColumn" = \'MyValue\'');
>         // You should use prepared_query() if using a dynamic value to prevent SQL-injection style attacks,
>         // though at the database level it's usually slower than the non-parameterised query()
>         $myValue = 'MyValue';
>         $result = DB::prepared_query('SELECT * FROM "MyTable" WHERE "MyColumn" = ?', [$myValue]);
>         foreach ($result as $row) {
>             // Process the row
>         }
>     }
> }
> ```

Reviewer: ```suggestion
Refer to the [raw SQL](/developer_guides/model/data_model_and_orm/#raw-sql) section for details about how to make raw SQL queries.
```

Author: Updated
